### PR TITLE
Prefer non-mutating PSPs, then order by name

### LINKED ISSUE #23217: Bring PodSecurityPolicy to usable state
The PodSecurityPolicy object was merged by @pweil- in #7893.

That defined a new type, (https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/v1beta1/types.go#L919), a registry, kubectl support, and a client.  

We still need to:
- [x] expose the extension by default, and generate the docs for it.
- [x] write code that decides whether a Pod create/update is allowed by any maching PSP. (assume this exists downstream in openshift.)
- [x] decide on a way to determine which type applies to which requests; that is, something to bind request attributes, like user, groups, namespace, with the PSPs that apply to that pod.
- [x] write admission controller code that does the previous two things and says yes or no to Pod creates/updates.
- [ ] define behavior when multiple PodSecurityPolicy objects allow a pod - #52849
- [ ] define a way for policy authors to affect which PSP is selected when multiple PodSecurityPolicy objects allow a pod - #52849

### LINKED ISSUE #36184: Split PSP defaulting and validation
When authz is added to PSP admission (https://github.com/kubernetes/kubernetes/pull/33080) the user submitting the request will now be taken into account for PSP selection.  That means that if one user creates a pod and another user updates the pod the PSP options that the pod validates against may be different.  

Since the PodSpec should not be mutable on the fields that PSP defaults anyway the PSP logic should be split so that defaulting only occurs during an ADD operations.  An UPDATE operation should only attempt to validate that the user has a PSP that would allow the requested spec.

ref: https://github.com/kubernetes/kubernetes/pull/33080/files#r84148684

### PR DESCRIPTION
**Pull request by liggitt**
Fixes #36184
Fixes #23217
Related to #23217

Removes unnecessary mutation of pods:
* Determines effective security context for pods using a wrapper containing the pod and container security context, rather than building/setting a combined struct on every admission
* Does not set `privileged:&false` on security contexts with `privileged:nil`
* Does not set `runAsNonRoot:&true` on security contexts that already have a non-nil, non-0 `runAsUser`
* Does not mutate/normalize container capabilities unless changes are required (missing  defaultAddCapabilities or requiredDropCapabilities)

Defines behavior when multiple PSP objects allow a pod:
* PSPs which allow the pod as-is (no defaulting/mutating) are preferred
* If the pod must be defaulted/mutated to be allowed, the first PSP (ordered by name) to allow the pod is selected
* During update operations, when mutations to pod specs are disallowed, only non-mutating PSPs are used to validate the pod

```release-note
PodSecurityPolicy: when multiple policies allow a submitted pod, priority is given to ones which do not require any fields in the pod spec to be defaulted. If the pod must be defaulted, the first policy (ordered by name) that allows the pod is used.
```

### CODE REVIEW

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (php-coder) [inline plugin/pkg/admission/security/podsecuritypolicy/admission.go]: Do we really need to hold an entire provider if we need only its name?

**Comment by liggitt** [comment]
cc @kubernetes/sig-auth-pr-reviews

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (php-coder) [inline plugin/pkg/admission/security/podsecuritypolicy/admission.go]: Could you add more logging for all these branches? Even if they will be on level 6, it's better to have them for easy debugging.

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (liggitt) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: the goal was to keep the existing code intact and move it out of the loop

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (deads2k) [inline plugin/pkg/admission/security/podsecuritypolicy/admission.go]: open issue or something for this?  Sorting on name leads people do weird things.

Comment (php-coder) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: It's already here: https://github.com/kubernetes/kubernetes/pull/45358

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (tallclair) [inline plugin/pkg/admission/security/podsecuritypolicy/admission.go]: It should be possible to tell whether a PSP is mutating without applying it. Maybe add a TODO here to optimize that in the future? I would prefer it NOT to go into this PR though.

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (tallclair) [inline plugin/pkg/admission/security/podsecuritypolicy/admission.go]: nit: include namespace

**Review by tallclair — COMMENTED** [review]
+1 For this change. I think this should be considered a cherry-pick candidate for 1.8 patch release, since this is fixing an open bug in a beta feature.

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (liggitt) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: definitely agree... started down that road with https://github.com/kubernetes/kubernetes/pull/50169 and it got... large. The preemptive copy here is less than ideal, but this was focused on the minimal change to get the external behavior we want

**Comment by tallclair** [comment]
@liggitt - Are you waiting for anything on this? If it's just a matter of finding the time, I'd be happy to help.

**Comment by liggitt** [comment]
Got distracted by a couple 1.8.0 fires. Plan to add tests today/tomorrow, then will be ready

**Review by pweil- — APPROVED** [review]
I like this.  I think it's a clever solution that also allows people to use less restrictive PSPs if they have access without having to manipulate the pod spec or sorting.

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (liggitt) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: added logging

**Comment by liggitt** [comment]
updated with tests, removed one unnecessary defaulter, included a commit fixing tolerating of non-pod-spec mutations done by GC as part of delete

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (tallclair) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: ^^

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (tallclair) [inline plugin/pkg/admission/security/podsecuritypolicy/admission.go]: As long as you're doing the copying here, you might as well remove the copying logic from `assignSecurityContext` (it does a bunch of stuff to make sure it doesn't modify the pod if it's invalid).

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (tallclair) [inline plugin/pkg/admission/security/podsecuritypolicy/admission.go]: This might be a premature optimization, but you can skip this check if `len(providers) == 1`

File: plugin/pkg/admission/security/podsecuritypolicy/admission_test.go
Comment (tallclair) [inline plugin/pkg/admission/security/podsecuritypolicy/admission_test.go]: `oldPod` looks unused?

File: plugin/pkg/admission/security/podsecuritypolicy/admission_test.go
Comment (tallclair) [inline plugin/pkg/admission/security/podsecuritypolicy/admission_test.go]: Can you rename this to `expectMutation`?

**Review by tallclair — COMMENTED** [review]
Thanks for working on this.

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (tallclair) [inline plugin/pkg/admission/security/podsecuritypolicy/admission.go]: I don't understand this comment. Why does this change require the power to delete? Also what are the implications of changing the owner? (Sorry, not that familiar with GC).

Comment (liggitt) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: ownerref and finalizer fields can require updates during the deletion lifecycle of an object. requiring the actors that update those fields to have PSP permissions in order to remove ownerrefs/finalizers can eternally block deletion of objects.

updates that change only ownerRefs can trigger an object being deleted, but do not affect anything related to the pod's execution or security context

updates that change finalizers can block/unblock an object from being deleted, but do not affect anything related to the pod's execution or security context

for that reason, we don't need to rerun PSP checks on update if only those fields are being touched.

**Comment by liggitt** [comment]
Realized while adding more tests that `Admit()` always sets a non-null securitycontext on the pod and containers, even for a fully privileged PSP that requires no restriction. That counts as a mutating change. Looking into ways to resolve that now.

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (liggitt) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: done (which greatly simplified that function), and added *lots* of tests to make sure pods that pass validation as-is are not mutated by the top-level Admit() call

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (liggitt) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: only if `specMutationAllowed` is true... I'd rather keep this straightforward and deal with optimizations when we profile where the hotspots are

File: plugin/pkg/admission/security/podsecuritypolicy/admission_test.go
Comment (liggitt) [inline plugin/pkg/admission/security/podsecuritypolicy/admission_test.go]: changed this to start with a permissive PSP and tighten just the SELinux aspect of it to isolate tests

File: plugin/pkg/admission/security/podsecuritypolicy/admission_test.go
Comment (liggitt) [inline plugin/pkg/admission/security/podsecuritypolicy/admission_test.go]: same here, wanted to test just the mustRunAs aspect

File: plugin/pkg/admission/security/podsecuritypolicy/admission_test.go
Comment (liggitt) [inline plugin/pkg/admission/security/podsecuritypolicy/admission_test.go]: changing this to be otherwise unrestricted lets us ensure that the RunAsAny strategies don't mutate pods

File: plugin/pkg/admission/security/podsecuritypolicy/admission_test.go
Comment (liggitt) [reply plugin/pkg/admission/security/podsecuritypolicy/admission_test.go]: fixed

File: plugin/pkg/admission/security/podsecuritypolicy/admission_test.go
Comment (liggitt) [inline plugin/pkg/admission/security/podsecuritypolicy/admission_test.go]: removed since we pass copies of the pod to `assignSecurityContext`, rather than mutating and rolling back on failure

**Comment by tallclair** [comment]
Oh man, this is becoming quite large. I propose splitting it into 2 PRs:

1. Keep the alphabetical & non-mutating ordering, but change the logic to determine if a PSP is mutating. Rather than considering the (Pod)SecurityContext, _only_ look at the PSP and whether any mutating fields are set. Not as good, but good enough to make the PSP useable (and hopefully small enough to backport to 1.8). E.g. a PSP is "mutating" if it sets any of: RequiredDropCaps, DefaultAddCaps, AppArmor / Seccomp default profile, (Default)AllowPrivEsc, RO FS, etc.

2. In the second PR, include all the logic you have here to reduce unnecessary mutations, and compare with the actual pod. This second PR probably won't be cherrypicked to 1.8.

WYDT?

**Comment by liggitt** [comment]
> Oh man, this is becoming quite large.

Agree, though I think the fixes to avoid unnecessary mutations are required to make the selection actually work as described. 

> a PSP is "mutating" if it sets any of: RequiredDropCaps, DefaultAddCaps, AppArmor / Seccomp default profile, (Default)AllowPrivEsc, RO FS, etc.

There's even more that can mutate a pod. MustRunAs user strategy defaults. MustRunAs selinux strategy defaults. Etc.

Even just passing a pod through PSP admission mutates it due to denormalization of effective security contexts into the containers (which is what the biggest change here fixes)

File: pkg/security/podsecuritypolicy/provider.go
Comment (pweil-) [inline pkg/security/podsecuritypolicy/provider.go]: what is the purpose for this?  I think we lose the intention/protection of the setting via the kubelet if we're conditionally setting it.  I guess the argument is that if something later sets `runAsUser` then it could also set `runAsNonRoot` to false so it really isn't protection anyway?

**Comment by pweil-** [comment]
> denormalization of effective security contexts into the containers (which is what the biggest change here fixes)

Just to confirm this, the output is still the effective security context on all components?  The reasoning was that we didn't want to require anything looking at a SecurityContext to have to remember to check the effective settings vs the actual settings.

**Comment by liggitt** [comment]
> The reasoning was that we didn't want to require anything looking at a SecurityContext to have to remember to check the effective settings vs the actual settings.

No. All components still have to do that, since PSP is optional (the kubelet computes the effective security context as well).

File: pkg/security/podsecuritypolicy/provider.go
Comment (liggitt) [reply pkg/security/podsecuritypolicy/provider.go]: Today, a `MustRunAsNonRoot` PSP would mutate a pod where all the containers had `runAsNonRoot:null, runAsUser:5`, which is unnecessary (they all are explicitly not running as root already), breaks update cases (since validation would prevent mutation of the spec), and breaks detection of which PSPs would allow the pod as-is

Comment (liggitt) [reply pkg/security/podsecuritypolicy/provider.go]: If the container already sets an explicit runAsUser that is non-zero, there's no kubelet-side protection we need

**Comment by pweil-** [comment]
> No. All components still have to do that, since PSP is optional (the kubelet computes the effective security context as well).

Gah, yes...you can't count on this admission controller ever being there :sob: Need to make sure everything is using the effective context then.

File: pkg/registry/rbac/helpers.go
Comment (enj) [inline pkg/registry/rbac/helpers.go]: We need the reflect based nil check?

Comment (liggitt) [reply pkg/registry/rbac/helpers.go]: yes. some of the tests pass nil -> runtime.Object -> NewAdmissionAttribute() which ends up as a nil-referencing interface

**Comment by tallclair** [comment]
> There's even more that can mutate a pod. MustRunAs user strategy defaults. MustRunAs selinux strategy defaults. Etc.

The most important (and probably commonly used) restrictions are `Privileged` and `Host{Path,Network,IPC,PID}`. If we could get just those pieces working in 1.8 I would be happy, and fortunately none of those are mutating (except privileged, which is a small fix).

> Even just passing a pod through PSP admission mutates it due to denormalization of effective security contexts into the containers (which is what the biggest change here fixes)

Can we just init empty secruity contexts at the beginning of the PSP controller for 1.8? I know it's not ideal, but the PSP is already doing that implicitly, so it's not actually a change from the current behavior.

```go
// Init empty security contexts if none are set prior to checking PodSecurityPolicies
if a.GetOperation() == admission.Create {
  if pod.SecurityContext == nil {
    pod.SecurityContext = &api.PodSecurityContext{}
  }
  for _, c := range append([]api.Container, pod.Containers, pod.InitContainers) {
    if c.SecurityContext == nil {
      c.SecurityContext = &api.SecrityContext{}
    }
  }
}
```

To clarify, I'm not dismissing the changes you've made in this PR. I think these are good improvements and we should get them into 1.9. However, I'm proposing these intermediate (hacky) steps as a compromise for the sake of making the PSP useable in 1.8.

File: pkg/securitycontext/accessors.go
Comment (tallclair) [inline pkg/securitycontext/accessors.go]: I think this check should be order agnostic?

File: pkg/securitycontext/accessors.go
Comment (tallclair) [inline pkg/securitycontext/accessors.go]: nit: what's the purpose of this check?

**Comment by liggitt** [comment]
> Can we just init empty secruity contexts at the beginning of the PSP controller for 1.8?

(I think) it's less straightforward than that... it would be updating each container with its computed effective security context prior to checking against policies and determining mutation

> I'm proposing these intermediate (hacky) steps as a compromise for the sake of making the PSP useable in 1.8.

my concern is that with a partial fix here we'll replace the "known undefined" state of multiple matching policies with "defined, but incorrect" state, and that the full fix would change behavior in ways that would break people.

File: pkg/security/podsecuritypolicy/provider.go
Comment (liggitt) [reply pkg/security/podsecuritypolicy/provider.go]: yeah, that got way cleaner

File: pkg/security/podsecuritypolicy/provider.go
Comment (liggitt) [reply pkg/security/podsecuritypolicy/provider.go]: done

File: pkg/security/podsecuritypolicy/capabilities/mustrunas.go
Comment (liggitt) [reply pkg/security/podsecuritypolicy/capabilities/mustrunas.go]: strategies that take pod attributes (like namespace) into account are planned, so keeping the interfaces as is makes sense, we just didn't want each validation to have to recompute effective capabilities

File: pkg/security/podsecuritypolicy/group/runasany.go
Comment (liggitt) [reply pkg/security/podsecuritypolicy/group/runasany.go]: not all strategies make use of pod attributes, but some could

File: pkg/securitycontext/accessors.go
Comment (liggitt) [reply pkg/securitycontext/accessors.go]: for all of these, before changing the container security context from nil to non-nil, make sure we have to, by checking if the new value is different than the old effective value

File: pkg/security/podsecuritypolicy/group/runasany.go
Comment (liggitt) [reply pkg/security/podsecuritypolicy/group/runasany.go]: sure

File: pkg/securitycontext/accessors.go
Comment (liggitt) [reply pkg/securitycontext/accessors.go]: fixed

**Comment by tallclair** [comment]
BTW - if you want to test these changes in the e2e environment, you can try patching in https://github.com/kubernetes/kubernetes/pull/53476 (and actually restoring the restricted profile from https://github.com/kubernetes/kubernetes/pull/52367 should work with these changes)

File: pkg/securitycontext/accessors.go
Comment (liggitt) [reply pkg/securitycontext/accessors.go]: another thread, another commit. fixed the over-enthusiastic normalization of capabilities in `capabilities/mustrunas.go`. I'm ambivalent about re-normalizing for comparison here

File: plugin/pkg/admission/security/podsecuritypolicy/admission_test.go
Comment (php-coder) [inline plugin/pkg/admission/security/podsecuritypolicy/admission_test.go]: Why you didn't create a function that will create a pod with the desired properties instead of doing DeepCopy each time? What are the benefits?

File: plugin/pkg/admission/security/podsecuritypolicy/admission_test.go
Comment (php-coder) [inline plugin/pkg/admission/security/podsecuritypolicy/admission_test.go]: So, this commit also contains (a bit modified version of) https://github.com/kubernetes/kubernetes/pull/48126 ?

File: plugin/pkg/admission/security/podsecuritypolicy/admission_test.go
Comment (liggitt) [reply plugin/pkg/admission/security/podsecuritypolicy/admission_test.go]: no strong preference either way

File: pkg/securitycontext/accessors.go
Comment (php-coder) [reply pkg/securitycontext/accessors.go]: I didn't find where/how it was fixed. @liggitt did you push these changes or what am I missed here?

Comment (liggitt) [reply pkg/securitycontext/accessors.go]: oops, did all the container ones, missed this one

**Comment by liggitt** [comment]
TestCRD flake
/retest

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (tallclair) [inline plugin/pkg/admission/security/podsecuritypolicy/admission.go]: (aside) What is the purpose of this annotation? Specifically, what is the desired behavior on updates?

**Review by tallclair — COMMENTED** [review]
/lgtm

Thanks for tackling this mess.

File: plugin/pkg/admission/security/podsecuritypolicy/admission.go
Comment (liggitt) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: informational, to surface which PSP the most recent mutation was allowed under

Comment (tallclair) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: > ... most recent mutation was allowed under

In that case, should we not set this if `!specMutationAllowed`?

Comment (liggitt) [reply plugin/pkg/admission/security/podsecuritypolicy/admission.go]: `specMutationAllowed` just means the *PSP* is not allowed to change the spec (other things can, changing `image`, for example). It's still processing a pod Update, so I'd expect it to record which PSP allowed the update.

**Comment by liggitt** [comment]
added unit tests for accessors, no other changes

**Comment by liggitt** [comment]
look like the following packages have no OWNERS files (both should probably be some combination of sig-node and sig-auth reviewers/approvers)
* pkg/security/podsecuritypolicy
* pkg/securitycontext

**Comment by liggitt** [comment]
@derekwaynecarr for admission approval

**Comment by liggitt** [comment]
/assign @dchen1107 @smarterclayton
for pkg approval (pkg/security/podsecuritypolicy and pkg/securitycontext changes)

**Comment by smarterclayton** [comment]
/approve no-issue

Looks good as well to me

**Comment by igalarzab** [comment]
Thank you for the PR! Are there any plans on cherry-picking this for the v1.8.x branch as @tallclair mentioned earlier?

**Comment by liggitt** [comment]
Pick is open in https://github.com/kubernetes/kubernetes/issues/54029 pending review and soak time

File: pkg/securitycontext/accessors.go
Comment (php-coder) [inline pkg/securitycontext/accessors.go]: @liggitt Why do we need `ContainerSecurityContextMutator` here? It looks weird to require mutate for an accessor.

Comment (liggitt) [reply pkg/securitycontext/accessors.go]: just because it was reusing the concrete type which required it. I wouldn't object to taking an accessor and wrapping in a no-op mutator impl

**Pull request merged**
